### DISCUSSION TITLE
filename, directory and baseDirectory with iOS Image Picker

### DISCUSSION THREAD

**Discussion by MartinSchultz** [781flyingdutchman/background_downloader#150] [General]
First of all, thanks for the great plugin!

I fail to find the right combination of what to set for filename, directory and baseDirectory if I want to upload a photo from the library. I get from the asset picker back (on iOS) something like this:

/Users/martin/Library/Developer/CoreSimulator/Devices/40394A4F-1D64-4704-9F0B-A62E32514FA5/data/Containers/Data/Application/946D9300-FC20-42FD-B600-3A0E92376AC3/tmp/flutter-images/27f440039fee318ad8d0262a052075a4_exif.jpg

I set so far filename = 27f440039fee318ad8d0262a052075a4_exif.jpg
and getting:
Unhandled Exception: Invalid argument(s): Directory must be relative to the baseDirectory specified in the baseDirectory argument

directory: directory,
 baseDirectory: BaseDirectory.applicationSupport,

directory is the full path minus the filename. 

Any idea? Thx :)

**Comment by 781flyingdutchman** [comment]
Hi, thanks for reaching out. I am not familiar with the asset picker on iOS, but from the directory path you list it seems like those assets are stored in a separate (temporary) location. The `BaseDirectory` entries in the downloader lead to directories that end like this:
* 946D9300-FC20-42FD-B600-3A0E92376AC3/Documents for the `.applicationDocuments` directory
* 946D9300-FC20-42FD-B600-3A0E92376AC3/Library/Caches for the `.temporary` directory
* 946D9300-FC20-42FD-B600-3A0E92376AC3/Library/Application Support for the `.applicationSupport` directory
* 946D9300-FC20-42FD-B600-3A0E92376AC3/Library for the `.applicationLibrary` directory

946D9300-FC20-42FD-B600-3A0E92376AC3 is the dynamic app identifier, which will change every time the app is started. This is the reason why we can't just use a full path, as that path may no longer exist by the time the downloader starts its background task.

As you can see, neither of these directories resolve to something that follows the dynamic app identifier with /tmp, so effectively that directory is not reachable by the downloader. The solution is to copy any files you want to upload from the directory where you find it to one of the directories supported by the downloader, as referenced by the `BaseDirectory` enum. Those directory paths can be obtained using the `path_provider` package (the names are the same).  Alternatively, if you can instruct the asset picker to load the files to a regular directory (e.g. Application Support) then you don't have to copy it.

The full path that the downloader file location resolves to is the directory that `baseDirectory` points to, plus any directory you may have added as `directory`, plus the filename, where "plus" means concatenated using the path separator "/".

Hope this helps.

**Reply by MartinSchultz** [reply]
This helps a lot! :) Thanks
